[ENH] Add Image Overlay: new widget

ngergihun opened the pull request:
## New widget: OWOverlay

### Functionality: 

The user can choose a feature from the "Overlay Data" input, which is added to the attributes of the "Data" input table as a "visible image". This way, it is possible to add a visible image overlay even if the reader cannot read it from the original data file, or if it is stored separately. 

The widget creates an RGB overlay image from the displayed image. Thus, changing the display parameters directly affects the assigned visible overlay image.

### Outstanding issues:
- [ ] New icon
- [x] Weird test problem

One of the tests does not seem to run correctly, even though in a real application, the same thing works. Somehow, during `test_add_visible_image`, the image acquired from the imageplot is None, because `im = self.imageplot.img.qimage` in `set_vimage` of the OWOverlay. However, everything runs correctly when the widget is used. 

@markotoplak : Could you please assist us with this? What could cause this weird issue with the tests?

borondics commented:
The functionality is great for me. This is what we discussed with @markotoplak.

Here is an icon suggestion.

![OWOverlay](https://github.com/user-attachments/assets/1725d807-1320-40fc-800d-16ea195e050b)

borondics commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,281 @@
+
+
+class OWOverlay(OWWidget):
+    name = "Add Visible"
```
Yes, we need to make this general. "Add Image Overlay" works for me.

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,281 @@
+                
+                # Assign it to the datatable attributes
+                if self.maindata and vimage is not None:
+                    self.maindata.attributes['visible_images'] = [vimage]
```
We should NEVER modify the input data in any way. We really have to ensure that the widgets only work forward, not edit the input data.

markotoplak replied:
Here, copying the data, and (maybe also) that dictionary, is needed. Also, there should be a test that ensures that input data dict did not change at all.

markotoplak replied:
Also, it is wrong that visible images are replaced (this one should be added).

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,281 @@
+
+                # Need to modify the position and the scale since visual imageplot 
+                # place the corner of the pixel to the given position
+                vimage = ConstantBytesVisibleImage(name="External Image",
```
Because images should be added, an unused name should be chosen here (Orange has functions for choosing new names)

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,281 @@
+        self.imageplot.update_view()
+        self.commit.deferred()
+    
+    def set_vimage(self, data):
```
The reason that tests are failing is that this function is sloppy. This function should be asynchronous (threaded), because it relies on painting that is itself asynchronous. The check `if im is not None` should not be used for avoiding outputs if the image is not yet draw. Image should always exist here unless there was an error.

ngergihun commented:
@markotoplak Thanks for the suggestions. I tried to correct some parts, and the tests now seemed to run. I was unsure which native Orange functions you were referring to for choosing new names, so I've added some basic incremental overlay image name changes and included a test for this. This can be changed, of course.

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -246,9 +248,14 @@ def set_vimage(self, data):
-                    self.maindata.attributes["visible_images"] = [vimage]
+                # # Assign it to the datatable attributes
+                if newmaindata and vimage is not None:
+                    if "visible_images" in list(newmaindata.attributes):
```
`.attributes` is a dict, so conversion to list is not needed (and it actually makes it, O(n), while dict lookup is O(1); though that is not important here). So: `if "visible_images" in newmaindata.attributes:`

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,317 @@
+                        im.name for im in newmaindata.attributes["visible_images"]
+                    ]
+                basename = "Overlay Image"
+                name = get_unused_name(allnames, basename)
```
@ngergihun, `get_unique_names` from `Orange.data.util` :)

markotoplak commented on orangecontrib/spectroscopy/widgets/owoverlay.py:
```diff
@@ -0,0 +1,317 @@
+
+                # Update name
+                allnames = []
+                if "visible_images" in list(newmaindata.attributes):
```
Don't convert to `list`

markotoplak commented:
> Here is an icon suggestion.

@borondics, I like the look of icon, but apart from in the browser it does not render well for me. I could also not open it in inkspace nicely to reconvert (it shows some gibberish).

ngergihun commented:
@markotoplak Thanks for the reviews and suggestions. I changed to `get_unique_names` and corrected my mistake with the list conversion. Originally I wanted to write `list(newmaindata.attributes.keys())` but just `newmaindata.attributes` is much nicer, thanks.

markotoplak commented:
I think the new icon is wonderful. Thanks!
